core/hooks/pcapng.cc

```diff
@@ -0,0 +1,254 @@
+// Copyright (c) 2014-2016, The Regents of the University of California.
+// Copyright (c) 2016-2017, Nefeli Networks, Inc.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice, this
+// list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimer in the documentation
+// and/or other materials provided with the distribution.
+//
+// * Neither the names of the copyright holders nor the names of their
+// contributors may be used to endorse or promote products derived from this
+// software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include "pcapng.h"
+
+#include <fcntl.h>
+#include <sys/uio.h>
+#include <unistd.h>
+
+#include <limits>
+
+#include <glog/logging.h>
+
+#include "../message.h"
+#include "../utils/common.h"
+#include "../utils/pcapng.h"
+#include "../utils/time.h"
+
+using namespace bess::utils::pcapng;
+
+namespace {
+
+// Return `a` rounded up to the nearest multiple of `b`
+template <typename T>
+T RoundUp(T a, T b) {
+  return ((a + (b - 1)) / b) * b;
+}
+
+// Return the number of bytes of padding to align a buffer long `a` to
+// `b` units.
+template <typename T>
+T PadSize(T a, T b) {
+  return RoundUp(a, b) - a;
+}
+
+// Return the single hex digit representing `nibble`.  If it cannot be
+// represented, return the char 'X'.
+char NibbleToHD(char nibble) {
+  if (nibble >= 0 && nibble <= 9) {
+    return nibble + '0';
+  } else if (nibble >= 10 && nibble <= 15) {
+    return nibble - 10 + 'A';
+  } else {
+    return 'X';
+  }
+}
+
+// Represent the buffer `src` (long `len` bytes) as an hex string.  Put
+// the result into `dst` (which must be at least `len * 2` bytes long).
```

Comment:
It seems like there ought to be some library function for this (especially, a fancy version like `hexdump -C` output), but I don't know of any... oh well, this will do fine.